File: Leetcode/Python/2658. Maximum Number of Fish in a Grid/solution.py
# https://leetcode.com/problems/maximum-number-of-fish-in-a-grid/description

class Solution:
    def findMaxFish(self, grid: List[List[int]]) -> int:
        dy = [0, 0, 1, -1]
        dx = [1, -1, 0, 0]
        max_fish = 0

        for y in range(len(grid)):
            for x in range(len(grid[0])):
                if grid[y][x] != 0:
                    total_fish = 0
                    stack = [(x, y)]

                    while stack:
                        cur_x, cur_y = stack.pop()

                        total_fish += grid[cur_y][cur_x]
                        grid[cur_y][cur_x] = 0

                        for idx in range(4):
                            next_x, next_y = cur_x + dx[idx], cur_y + dy[idx]

                            if 0 <= next_x < len(grid[0]) and 0 <= next_y < len(grid) and grid[next_y][next_x] != 0:
                                stack.append((next_x, next_y))

                    max_fish = max(max_fish, total_fish)

        return max_fish

File: Leetcode/Typescript/2658. Maximum Number of Fish in a Grid/solution.ts
// https://leetcode.com/problems/maximum-number-of-fish-in-a-grid/description

function findMaxFish(grid: number[][]): number {
  const dx = [0, 0, 1, -1];
  const dy = [1, -1, 0, 0];
  let maxFish = 0;

  for (let y = 0; y < grid.length; y++) {
    for (let x = 0; x < grid[0].length; x++) {
      if (grid[y][x] !== 0) {
        let totalFish = 0;
        const stack: number[][] = [];
        stack.push([x, y]);

        while (stack.length) {
          const [curX, curY] = stack.pop();

          totalFish += grid[curY][curX];
          grid[curY][curX] = 0;

          for (let idx = 0; idx < 4; idx++) {
            const nextX = curX + dx[idx],
              nextY = curY + dy[idx];

            if (
              0 <= nextX &&
              nextX < grid[0].length &&
              0 <= nextY &&
              nextY < grid.length &&
              grid[nextY][nextX] !== 0
            ) {
              stack.push([nextX, nextY]);
            }
          }
        }

        maxFish = Math.max(maxFish, totalFish);
      }
    }
  }

  return maxFish;
}
